src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DocumentHighlightHandler.cs

```diff
@@ -0,0 +1,127 @@
+        public DocumentHighlightHandler(
+            LSPRequestInvoker requestInvoker,
+            LSPDocumentManager documentManager,
+            LSPProjectionProvider projectionProvider,
+            LSPDocumentMappingProvider documentMappingProvider)
+        {
+            if (requestInvoker is null)
+            {
+                throw new ArgumentNullException(nameof(requestInvoker));
+            }
+
+            if (documentManager is null)
+            {
+                throw new ArgumentNullException(nameof(documentManager));
+            }
+
+            if (projectionProvider is null)
+            {
+                throw new ArgumentNullException(nameof(projectionProvider));
+            }
+
+            if (documentMappingProvider is null)
+            {
+                throw new ArgumentNullException(nameof(documentMappingProvider));
+            }
+
+            _requestInvoker = requestInvoker;
+            _documentManager = documentManager;
+            _projectionProvider = projectionProvider;
+            _documentMappingProvider = documentMappingProvider;
+        }
+
+        public async Task<DocumentHighlight[]> HandleRequestAsync(DocumentHighlightParams request, ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            if (clientCapabilities is null)
+            {
+                throw new ArgumentNullException(nameof(clientCapabilities));
+            }
+
+            if (!_documentManager.TryGetDocument(request.TextDocument.Uri, out var documentSnapshot))
+            {
+                return null;
+            }
+
+            var projectionResult = await _projectionProvider.GetProjectionAsync(documentSnapshot, request.Position, cancellationToken).ConfigureAwait(false);
+            if (projectionResult == null || projectionResult.LanguageKind != RazorLanguageKind.CSharp)
+            {
+                return null;
+            }
+
+            var serverKind = LanguageServerKind.CSharp;
+            var documentHighlightParams = new DocumentHighlightParams()
+            {
+                Position = projectionResult.Position,
+                TextDocument = new TextDocumentIdentifier()
+                {
+                    Uri = projectionResult.Uri
+                }
+            };
+
+            var result = await _requestInvoker.ReinvokeRequestOnServerAsync<DocumentHighlightParams, DocumentHighlight[]>(
+                Methods.TextDocumentDocumentHighlightName,
+                serverKind,
+                documentHighlightParams,
+                cancellationToken).ConfigureAwait(false);
+
+            if (result == null || result.Length == 0)
+            {
+                return result;
+            }
+
+            var remappedHighlights = new List<DocumentHighlight>();
+
+            foreach (var highlight in result)
```

Comment:
Food for thought: Almost every location of MapToDocumentRangeAsync has this type of pattern where we iterate over a  list of items and then map each one individually. We should change this API to reflect that typical usage to avoid going over the wire every time. I'm imagining a large C# document with multiple usages and the time spent will be hugeeee with how things currently work

Reply:
https://github.com/dotnet/aspnetcore/issues/22289